homeassistant/components/sense/config_flow.py

```diff
@@ -21,37 +25,74 @@
+    async def validate_input(self, data):
+        """Validate the user input allows us to connect.
 
+        Data has the keys from DATA_SCHEMA with values provided by the user.
+        """
+        self._auth_data.update(dict(data))
+        timeout = self._auth_data[CONF_TIMEOUT]
+        client_session = async_get_clientsession(self.hass)
 
-class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
-    """Handle a config flow for Sense."""
+        self._gateway = ASyncSenseable(
+            api_timeout=timeout, wss_timeout=timeout, client_session=client_session
+        )
+        self._gateway.rate_limit = ACTIVE_UPDATE_RATE
+        await self._gateway.authenticate(
+            self._auth_data[CONF_EMAIL], self._auth_data[CONF_PASSWORD]
+        )
 
-    VERSION = 1
+    async def create_entry_from_data(self):
+        """Create the entry from the config data."""
+        self._auth_data["access_token"] = self._gateway.sense_access_token
+        self._auth_data["user_id"] = self._gateway.sense_user_id
+        self._auth_data["monitor_id"] = self._gateway.sense_monitor_id
+        existing_entry = await self.async_set_unique_id(self._auth_data[CONF_EMAIL])
+        if not existing_entry:
+            return self.async_create_entry(
+                title=self._auth_data[CONF_EMAIL], data=self._auth_data
+            )
 
-    async def async_step_user(self, user_input=None):
-        """Handle the initial step."""
+        self.hass.config_entries.async_update_entry(
+            existing_entry, data=self._auth_data
+        )
+        await self.hass.config_entries.async_reload(existing_entry.entry_id)
+        return self.async_abort(reason="reauth_successful")
+
+    async def validate_input_and_create_entry(self, user_input, errors):
+        """Validate the input and create the entry from the data."""
+        try:
+            await self.validate_input(user_input)
+            return await self.create_entry_from_data()
+        except SenseMFARequiredException:
+            return await self.async_step_validation()
+        except SENSE_TIMEOUT_EXCEPTIONS:
+            errors["base"] = "cannot_connect"
+        except SenseAuthenticationException:
+            errors["base"] = "invalid_auth"
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = "unknown"
+        return None
+
+    async def async_step_validation(self, user_input=None):
+        """Handle validation (2fa) step."""
         errors = {}
-        if user_input is not None:
+        if user_input:
             try:
-                info = await validate_input(self.hass, user_input)
-                await self.async_set_unique_id(user_input[CONF_EMAIL])
-                return self.async_create_entry(title=info["title"], data=user_input)
+                await self._gateway.validate_mfa(user_input[CONF_CODE])
+                return await self.create_entry_from_data()
```

Comment:
Existing but since it's changed, please move this into an else block since it's not excepted to throw any of the exceptions